cmd/yggd/client.go

```diff
@@ -11,41 +11,123 @@ import (
+	}
+}
+
+// Connect starts a goroutine receiving values from the client's dispatcher and
+// transmits the data using the transporter.
 func (c *Client) Connect() error {
+	if c.t == nil {
+		return fmt.Errorf("cannot connect client: missing transport")
+	}
+
+	// Connect the Dispatcher
+	if err := c.d.Connect(); err != nil {
+		return fmt.Errorf("cannot connect dispatcher: %w", err)
+	}
+
+	// start receiving values from the dispatcher and transmit them using the
+	// provided transporter.
+	go func() {
+		for msg := range c.d.Outbound {
+			code, metadata, data, err := c.SendDataMessage(&msg.Data, msg.Data.Metadata)
+			if err != nil {
+				log.Errorf("cannot send data message: %v", err)
+				continue
+			}
+			msg.Resp <- struct {
+				Code     int
+				Metadata map[string]string
+				Data     []byte
+			}{
+				Code:     code,
+				Metadata: metadata,
+				Data:     data,
+			}
+		}
+	}()
+
+	// set a transport RxHandlerFunc that calls the client's control and data
+	// receive handler functions.
+	err := c.t.SetRxHandler(func(addr string, metadata map[string]interface{}, data []byte) error {
+		switch addr {
+		case "data":
+			var message yggdrasil.Data
+
+			if err := json.Unmarshal(data, &message); err != nil {
+				return fmt.Errorf("cannot unmarshal data message: %w", err)
+			}
+			if err := c.ReceiveDataMessage(&message); err != nil {
+				return fmt.Errorf("cannot process data message: %w", err)
+			}
+		case "control":
+			var message yggdrasil.Control
+
+			if err := json.Unmarshal(data, &message); err != nil {
+				return fmt.Errorf("cannot unmarshal control message: %w", err)
+			}
+			if err := c.ReceiveControlMessage(&message); err != nil {
+				return fmt.Errorf("cannot process control message: %w", err)
+			}
+		default:
+			return fmt.Errorf("unsupported destination type: %v", addr)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("cannot set RxHandler: %v", err)
+	}
+
 	return c.t.Connect()
 }
 
-func (c *Client) SendDataMessage(msg *yggdrasil.Data) ([]byte, error) {
-	return c.sendMessage(msg, "data")
+func (c *Client) SendDataMessage(msg *yggdrasil.Data, metadata map[string]string) (int, map[string]string, []byte, error) {
+	return c.sendMessage("data", metadata, msg)
 }
 
-func (c *Client) SendConnectionStatusMessage(msg *yggdrasil.ConnectionStatus) ([]byte, error) {
-	return c.sendMessage(msg, "control")
+func (c *Client) SendConnectionStatusMessage(msg *yggdrasil.ConnectionStatus) (int, map[string]string, []byte, error) {
+	code, metadata, data, err := c.sendMessage("control", nil, msg)
+	if err != nil {
+		return -1, nil, nil, err
+	}
+	return code, metadata, data, nil
 }
 
-func (c *Client) SendEventMessage(msg *yggdrasil.Event) ([]byte, error) {
-	return c.sendMessage(msg, "control")
+func (c *Client) SendEventMessage(msg *yggdrasil.Event) (int, map[string]string, []byte, error) {
+	code, metadata, data, err := c.sendMessage("control", nil, msg)
+	if err != nil {
+		return -1, nil, nil, err
```

Comment:
I'm not sure what this  -1 is, could you add a const and return a const that it's easy to read and reduce the cognitive load.